### ISSUE TITLE
Add ability to use pundit_role in the same context as pundit_policy_class for owned fields. 

### ISSUE THREAD

**Issue by Austio** [rmosolgo/graphql-ruby#3387]
[Gist Demonstrating this Feature Report](https://gist.github.com/Austio/5c69436a0cb3b141f072d5bde4193c2e) the TODO(rob) are the spots that are important for understanding.

**Is your feature request related to a problem? Please describe.**

This feature would unify the usage of `pundit_policy_class` and `pundit_role` when using the Pundit Integration.

The change would be: when looking up the `pundit_role` we would add a check for `@owners` in the exact same way we currently do for `pundit_policy_class`

As part of the current flow for pundit we are able to declare 2 methods in Fields and Types `pundit_policy_class` and `pundit_role`
- pundit_policy_class will look for the "@owner" of a field when looking up the class
- pundit_role does not use "@owner" for looking up pundit_role

In an example that demonstrates the impact
- given you have a BaseField defining a global `pundit_role` of :general_global_access and a `pundit_policy_class` of MyAwesomePolicy 
  - and a type (User) that opts for another `pundit_policy_class` of UserPolicy and `pundit_role` of :read 
- then authorization will use the @owner User type's `pundit_policy_class` of UserPolicy with the global BaseField's `pundit_role` of :general_global_access

From an app flow standpoint, it would be better if I could declare a pundit_role and then have that role applied to the owned fields in the same way we currently handle this for the pundit class.  This would prevent scenarios such as this one

1. Super Secure By Default: Declaring a base field policy_role that is restrictive by default (say :never).  Doing this now is not ideal because you would have to allow :never on the Policies that are used on fields (in the scenario above, remember we will use the global pundit_role but the @owner pundit_policy_class)

It seems like this @owner `pundit_policy_class` was implemented as [part of a bug fix](https://github.com/rmosolgo/graphql-ruby/blob/master/CHANGELOG-pro.md#1912-22-feb-2019) so my hopes and dreams are that the `pundit_role` consideration is just something that we missed 

**Describe the solution you'd like**

Add a check for "@owner" when resolving the `pundit_role` 

**Describe alternatives you've considered**

The alternative is that I would have to monkey patch `pundit_role` so that it checks the owner of the field in the same way that `pundit_policy_class` is checking.

**Additional context**

[This issue was reported here previously.](https://github.com/rmosolgo/graphql-ruby/issues/2164#issuecomment-469706348)

**Comment by Austio** [comment]
@rmosolgo any thoughts on direction for including the pundit_role in a future release?  

Our course to move past this will be to monkey patch and would like to understand the scope of that decisions (is this something we will go it alone forever or just temp until the method is patched to match the pundit_policy_class lookup algo)

**Comment by rmosolgo** [comment]
Derp, thanks for the ping on this -- sorry it slipped off my radar. I agree it makes sense to make these lookup algorithms match. Let me take a look at how it can be integrated into the gem without disrupting anyone who might be using the existing behavior!

**Comment by Austio** [comment]
Sweet, thanks for the update and let me know if you need anything on this end.  I think the change will be this, if you get deep and it is going to vary please ping :)

```
    def pundit_role(new_role = READING_ROLE)
      if new_role != READING_ROLE
        @pundit_role = new_role
      elsif defined?(@pundit_role)
        @pundit_role
      elsif self.is_a?(Class) && superclass.respond_to?(:pundit_role)
        superclass.pundit_role
      # ------------- This conditional is the change to match the pundit_policy_class lookup
      elsif defined?(@owner) && @owner.respond_to?(:pundit_role)
         @owner.pundit_role
      elsif self.class.respond_to?(:pundit_role)
        self.class.pundit_role
      else
        false
      end
    end
 ```

**Comment by rmosolgo** [comment]
I'm sorry, after looking into it further, I'm not convinced that I can safely change the default behavior in the gem. The problem is, even if the change you've proposed here _does_ make more sense, it will cause behavioral changes to existing systems that are hard to grok and hard to straighten out. 

I'd like to understand the real-world problem a little better here, to see if there's a suitable restructuring that would satisfy the needs of the application while removing the need for a breaking change in the library. Looking at the attached gist, the required change is pretty small, basically, make `BasePolicyClass` the base class, instead of `UserPolicy`:

```diff
~/code/graphql-ruby % git diff --no-index test_before.rb test_after.rb
diff --git a/test_before.rb b/test_after.rb
index f2c674406..12bc34bbb 100644
--- a/test_before.rb
+++ b/test_after.rb
@@ -68,25 +68,13 @@ class Schema < GraphQL::Schema
     end
   end

-  class UserPolicy
+  class BasePolicyClass
     def initialize(user, object)
       @user = user
       @object = object
     end

-    # A user object can only be viewed if the current user
-    # _is_ that same user
-    def view?
-      @user == @object[:handle]
-    end
-
-    # TODO(rob) uncomment this and the schema will execute
-    # def general?
-    #   false
-    # end
-  end
-
-  class BasePolicyClass < UserPolicy
     def view?
       false
     end
@@ -96,6 +84,14 @@ class Schema < GraphQL::Schema
     end
   end

+  class UserPolicy < BasePolicyClass
+    # A user object can only be viewed if the current user
+    # _is_ that same user
+    def view?
+      @user == @object[:handle]
+    end
+  end
+
   class BaseField < GraphQL::Schema::Field
     include GraphQL::Pro::PunditIntegration::FieldIntegration
     include PunditInspectionHooks
```

That way, the expected method (`#general?`) is always implemented and returns `false` by default. Then, objects that want to make their fields _visibile_ by default can re-implement that check on their policies to return `true` (or apply some kind of check). 

Would a restructuring like that work in your real-world case? If not, could you tell me more about it?

**Comment by Austio** [comment]
First I appreciate you looking into this.  Could you help me understand the scenario where you are wanting this functionality?  It could be that there is something that I am completely missing here.

Absent something I am missing, my initial thought is that I don't think that we would be able to make this work because I think it causes the authentication to call incorrect methods when we are checking an 'owned' type.  

The core issue to me here is that the lookup algorithm is different for the Class than the method that will be called on the class, unless there is an architectural reason for this then I think the best method is to make them congruent even if it is a breaking change and requires a 2.0 bump.

Without this change, imagine you are wanting to implement something like 'public' as the default pundit_role and on some Type you define a pundit_policy_class of `AdminPolicy` on and a pundit_role of `edit`.

When the subfields check the owner it will lookup the `AdminPolicy` *BUT* it will use the global `public` pundit_role on it, which is different than the requirements of `edit`.  Your declaration that you want a role of `edit` is not respected, you would have to implement a `public` method on AdminPolicy and make it aware of the full context of the request to properly authenticate (the query, the owner etc)

**Comment by rmosolgo** [comment]
> When the subfields check the owner it will lookup the `AdminPolicy` BUT it _will_ use the global `public` pundit_role on it, which is different than the requirements of `edit`.

That's right, I guess in _my_ mind, I have a similar model as with a REST endpoint: once you can see an _object_, then you can likely see all its fields. Sometimes that _isn't_ true, but in my experience, that's the exception, not the rule. So, when you apply `AdminPolicy#edit` to an object type, that policy will be checked whenever GraphQL-Ruby encounters an object of that type. Personally, I would have _expected_ the base policy to implement `def public; true; end`, so authorization on fields would be a no-op most times. 

The cases I can _imagine_ for providing custom field-level roles are: 

- Query root fields, since they're entry points for the schema and can't rely on authorization checks on `Query` 
- Particularly sensitive fields on an object, which, again, in my own experience, are unusual, but happen. 

In those cases, it would make sense to me to provide an override `pundit_role:` on the field definition. 

Maybe this would be possible, while covering those cases:

- Whenever a field has an _override_ `pundit_role:`, it uses the owner's `pundit_policy_class`;
- Whenever a field uses the default `pundit_role` from its class, it uses the field class's `pundit_policy_class`.

In my own mind, that makes _sense_ -- it covers the cases I described above -- even if it is a bit complicated. I'll explore the viability of it in the code, too. What do you think of it?

**Comment by Austio** [comment]
@rmosolgo I created another gist that demonstrates better what I am saying with the lookup logic being off from what I expected.
https://gist.github.com/Austio/fd2c53a9067845d90e39e0d8568c3d5c 

I added a comment for the changes to the code if we lookup `pundit_role` in the same way we do `pundit_policy_class`.

Where I am starting from with this issue is that I think that Types are different from Fields and their authorization logic should be different and independent.  Right now, they are dependent in a way that I did not expect when I was working with them.

The issue is that when you declare a pundit_role for ANY field, then the Policy for the underlying mounted types will have to implement the method that is being called on the field policy class.  In the code below, it means that the `UserType#pundit_policy_class` will have to respond truthily to field_logged_in. 

```
 class Query < BaseObject
    graphql_name "Query"
    pundit_policy_class QueryPolicy
    pundit_role :view

    # Due to adding pundit_role field_logged_in, the UserType has to also implement and respond to field_logged_in truthy
    # NOTE: I named in this way to clearly distinguish the logic 
    field :user, UserType, pundit_role: :field_logged_in, null: true
    def user
      OpenStruct.new(name: 'jdoe')
    end
  end
```

**Comment by Austio** [comment]
@rmosolgo wanted to circle back to this.  Would it be possible to implement a flag to opt into parity between the pundit_role and pundit_policy_class

I think that would be a single change


```
    def pundit_role(new_role = READING_ROLE)
      if new_role != READING_ROLE
        @pundit_role = new_role
      elsif defined?(@pundit_role)
        @pundit_role
      elsif self.is_a?(Class) && superclass.respond_to?(:pundit_role)
        superclass.pundit_role
      # ------------- This conditional is the change to match the pundit_policy_class lookup
      elsif CONFIG.pundit_partity_flag? && defined?(@owner) && @owner.respond_to?(:pundit_role)
         @owner.pundit_role
      elsif self.class.respond_to?(:pundit_role)
        self.class.pundit_role
      else
        false
      end
    end
  ```

**Comment by rmosolgo** [comment]
:+1: Yes, that seems fine. I just shipped graphql-pro 1.17.13 with an option for that: 

```ruby 
class BaseField < GraphQL::Schema::Field 
  include GraphQL::Pro::PunditIntegration::FieldIntegration 
  use_owner_role(true)
end 
```

With `use_owner_role(true)`, it will do a lookup like you suggested above. Thanks for your patience and for working with me on this! 

If other folks come looking for a solution like this, I'll consider a migration path for making this the default behavior.

**Comment by Austio** [comment]
Awesome, thanks @rmosolgo !
